pkg/agent/generate_secret_controller.go

```diff
@@ -0,0 +1,147 @@
+// }
+
+// SetupWithManager sets up the controller with the Manager.
+func (c *ExternalSecretController) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&operatorapiv1.Klusterlet{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 1}).
+		//WithEventFilter(ExternalSecretPredicateFunctions).
+		Complete(c)
+}
+
+// Reconcile updates the Hypershift addon status based on the Deployment status.
+func (c *ExternalSecretController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	c.log.Info(fmt.Sprintf("reconciling annotation for %s", req.Name))
+	defer c.log.Info(fmt.Sprintf("done reconciling annotation for %s", req.Name))
+
+	if !strings.Contains(req.Name, "klusterlet-") {
+		c.log.Info("klusterlet not from a managed cluster")
+		return ctrl.Result{}, nil //No need to error
+	}
+
+	//Get klusterlet
+	klusterletNamespaceNsn := types.NamespacedName{
+		Name:      req.Name,
+		Namespace: req.Namespace,
+	}
+	klusterlet := &operatorapiv1.Klusterlet{}
+	if err := c.spokeClient.Get(ctx, klusterletNamespaceNsn, klusterlet); err != nil {
+		c.log.Error(err, "Unable to get klusterlet")
+		return ctrl.Result{}, err
+	}
+
+	_, hostedClusterName, _ := strings.Cut(req.Name, "klusterlet-")
+
+	lo := &client.ListOptions{}
+	hostedClusters := &hyperv1beta1.HostedClusterList{}
+
+	// List the HostedCluster objects across all namespaces
+	if err := c.spokeClient.List(ctx, hostedClusters, lo); err != nil {
+		c.log.Error(err, "Unable to list hosted clusters in all namespaces")
+		return ctrl.Result{}, err
+	}
+
+	hostedClusterObj := &hyperv1beta1.HostedCluster{}
+	// Loop over the list of HostedCluster objects and find the one with the specified name
+	for index, hc := range hostedClusters.Items {
+		if hc.Name == hostedClusterName {
+			hostedClusterObj = &hostedClusters.Items[index]
+			break
+		}
+		if index == len(hostedClusters.Items) {
+			errh := errors.New("could not retrieve hosted cluster")
+			c.log.Error(errh, fmt.Sprintf("Unable to find hosted clusters with name %s", hostedClusterName))
+			return ctrl.Result{}, errh
+		}
+	}
+
+	if klusterlet.ObjectMeta.DeletionTimestamp.IsZero() {
+
+		// Add finalizer if it doesn't exist
+		if !controllerutil.ContainsFinalizer(klusterlet, klusterletAnnotationFinalizer) {
+			controllerutil.AddFinalizer(klusterlet, klusterletAnnotationFinalizer)
+			c.log.Info(fmt.Sprintf("Added finalizer to %s", klusterlet.Name))
+		}
+
+		// Add the annotation to the hostedcluster if it doesn't exist
+		// Create the annotation map if it doesn't exist
+		if hostedClusterObj.ObjectMeta.Annotations == nil {
+			hostedClusterObj.ObjectMeta.Annotations = make(map[string]string)
+		}
+		if _, ok := hostedClusterObj.ObjectMeta.Annotations[hcAnnotation]; !ok {
+			hostedClusterObj.Annotations[hcAnnotation] = "true"
+			c.log.Info(fmt.Sprintf("Annotated %s with %s", hostedClusterObj.Name, hcAnnotation))
+		}
+
+	} else {
+		//Remove annotation from hosted cluster
```

Comment:
If the HostedCluster is being deleted, does it matter whether we remove this annotation or not?

Reply:
The klusterlet is being deleted, not the hostedcluster.